src/common/ui/setting-text.vue

```diff
@@ -23,39 +32,70 @@ export default {
       type: Boolean,
       default: true,
     },
+    hasSave: {
+      type: Boolean,
+      default: true,
+    },
+    hasReset: Boolean,
   },
   data() {
     return {
       value: null,
       jsonValue: null,
       error: null,
+      canSave: null,
+      canReset: null,
     };
   },
   created() {
     const handle = this.json
       ? (value => JSON.stringify(value, null, '  '))
       // XXX compatible with old data format
       : (value => (Array.isArray(value) ? value.join('\n') : value || ''));
-    this.revoke = hookSetting(this.name, val => { this.value = handle(val); });
-    if (this.json) this.$watch('value', this.parseJson);
+    this.revoke = hookSetting(this.name, val => {
+      this.savedValue = val;
+      const text = handle(val);
+      if (this.value !== text) this.value = text;
+      else this.onInput(val);
+    });
+    this.defaultValue = objectGet(defaults, this.name);
   },
   beforeDestroy() {
     this.revoke();
   },
+  watch: {
+    value(val) {
+      if (this.json) {
+        try {
+          val = JSON.parse(val);
+          this.jsonValue = val;
+          this.error = null;
+        } catch (e) {
+          this.error = e.message || e;
+        }
+      }
+      this.onInput(val);
+    },
+  },
   methods: {
-    parseJson() {
-      try {
-        this.jsonValue = JSON.parse(this.value);
-        this.error = null;
-      } catch (e) {
-        this.error = e.message || e;
+    /** `event` is absent when calling from save() */
+    onChange(event) {
+      if (!this.error) {
+        const value = this.json ? this.jsonValue : this.value;
+        if (this.sync || !event) options.set(this.name, value);
+        if (event) this.$emit('change', value);
+        return true;
       }
     },
-    onChange() {
-      if (this.error) return;
-      const value = this.json ? this.jsonValue : this.value;
-      if (this.sync) options.set(this.name, value);
-      this.$emit('change', value);
+    onInput(val) {
+      this.canSave = this.hasSave && !this.error && !deepEqual(val, this.savedValue || '');
+      this.canReset = this.hasReset && !deepEqual(val, this.defaultValue || '');
+    },
+    onSave() {
+      if (this.onChange()) this.$emit('save');
```

Comment:
Why is `onChange` called here? I think data should have been updated when `save` is emitted.

Reply:
It's for deduplication. The only purpose of a dedicate onSave method is to inform the parent so it could do something. I guess I can extract the actual saving into a separate method to reduce confusion.

Reply:
BTW, saving is performed automatically when the text field loses focus, so there's no need for the save button strictly speaking, we could do without it by using an indicator for unsaved state and `beforeunload` listener to prevent the loss of data.

Reply:
Extraction example:
```js
    onChange() {
      this.commit({ change: true });
    },
    onInput(val) {
      this.canSave = this.hasSave && !this.error && !deepEqual(val, this.savedValue || '');
      this.canReset = this.hasReset && !deepEqual(val, this.defaultValue || '');
    },
    onSave() {
      if (this.commit({ save: true })) {
        this.$emit('save');
      }
    },
    onReset() {
      options.set(this.name, this.defaultValue);
    },
    commit({ change, save }) {
      if (!this.error) {
        const value = this.json ? this.jsonValue : this.value;
        if (this.sync || save) options.set(this.name, value);
        if (change) this.$emit('change', value);
        return true;
      }
    },
```

Reply:
We can update data on `input` and save/sync on `change`.
A simple way to handle `click` of save button is to wrap in `setTimeout` and avoid calling `change` manually.

Reply:
> BTW, saving is performed automatically when the text field loses focus, so there's no need for the save button strictly speaking, we could do without it by using an indicator for unsaved state and `beforeunload` listener to prevent the loss of data.

Good idea, maybe showing a different border color or some icon for unsaved state is enough.

Reply:
> update data on input

What data?

> setTimeout 

It can be argued that an indirect approach isn't better than a direct call but maybe I didn't understand what exactly you mean.

Reply:
`data` refers to `this.jsonValue` here, which is not updated in time when the textarea is focused and save button is clicked.

`setTimeout` can be replaced with `this.$nextTick` here, to wait until reactive data is updated. In this way it does not need to call something that should not be done by it.

```js
onSave() {
  this.$nextTick(() => {
     // do stuff
  });
}
```

Reply:
I've simplified the code by removing `sync`, see if you like it now. As for removing the save button, let's keep it so people don't get nervous :-) and anyway it's also useful as an obvious and understandable indicator of the modified state.

P.S. there was a problem in `master` fixed by 9fc8b6b2 - on browser startup (when "restore previous session" is selected in browser options) the Violentmonkey dashboard didn't initialize properly in the active tab.